File: go/ingest/audio_test.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

func TestAudioExtractor_Name(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{})
	if ext.Name() != "audio-whisper" {
		t.Fatalf("Name() = %q, want %q", ext.Name(), "audio-whisper")
	}
}

func TestAudioExtractor_ContentTypes(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{})
	types := ext.ContentTypes()

	expected := []string{
		"audio/mpeg",
		"audio/wav",
		"audio/x-wav",
		"audio/flac",
		"audio/mp4",
		"audio/ogg",
		"audio/aac",
		"audio/x-ms-wma",
		"audio/opus",
	}

	if len(types) != len(expected) {
		t.Fatalf("ContentTypes() length = %d, want %d", len(types), len(expected))
	}
	for i, ct := range types {
		if ct != expected[i] {
			t.Fatalf("ContentTypes()[%d] = %q, want %q", i, ct, expected[i])
		}
	}
}

func TestAudioExtractor_Capability(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{})
	cap := ext.Capability()

	if !cap.RequiresBinary {
		t.Fatal("expected RequiresBinary = true")
	}

	extensionSet := make(map[string]struct{}, len(cap.Extensions))
	for _, e := range cap.Extensions {
		extensionSet[e] = struct{}{}
	}
	for _, want := range []string{".mp3", ".wav", ".flac", ".m4a", ".ogg", ".aac"} {
		if _, ok := extensionSet[want]; !ok {
			t.Fatalf("expected extension %q in capability", want)
		}
	}

	mimeSet := make(map[string]struct{}, len(cap.MIMETypes))
	for _, m := range cap.MIMETypes {
		mimeSet[m] = struct{}{}
	}
	for _, want := range []string{"audio/mpeg", "audio/wav", "audio/mp4", "audio/ogg"} {
		if _, ok := mimeSet[want]; !ok {
			t.Fatalf("expected MIME type %q in capability", want)
		}
	}

	if len(cap.MagicBytes) == 0 {
		t.Fatal("expected non-empty MagicBytes")
	}
}

func TestAudioExtractor_Available_NoPython(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{
		PythonBinary: "/nonexistent/python3",
	})

	avail, err := ext.Available(context.Background())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if avail {
		t.Fatal("expected available=false when python binary does not exist")
	}
}

func TestAudioExtractor_Available_CachesResult(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{
		PythonBinary: "/nonexistent/python3",
	})

	// First call.
	avail1, _ := ext.Available(context.Background())
	// Second call should use cache.
	avail2, _ := ext.Available(context.Background())

	if avail1 != avail2 {
		t.Fatal("expected cached result to match")
	}
	if avail1 {
		t.Fatal("expected false for nonexistent binary")
	}
}

func TestAudioExtractor_Extract_EmptyInput(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{})
	result, err := ext.Extract(context.Background(), []byte{}, ExtractOptions{
		ContentType: "audio/wav",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Skipped {
		t.Fatal("expected skipped=true for empty input")
	}
	if result.Reason != "empty audio input" {
		t.Fatalf("unexpected reason: %q", result.Reason)
	}
}

func TestAudioExtractor_Extract_FileTooLarge(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{
		MaxFileSize: 100, // 100 bytes max
	})

	largeInput := make([]byte, 200)
	_, err := ext.Extract(context.Background(), largeInput, ExtractOptions{
		ContentType: "audio/wav",
	})
	if err == nil {
		t.Fatal("expected error for oversized file")
	}
	if !strings.Contains(err.Error(), "exceeds maximum") {
		t.Fatalf("unexpected error message: %v", err)
	}
}

func TestAudioExtractor_ComputeTimeout(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{
		TimeoutPerMin: 30 * time.Second,
		MinTimeout:    120 * time.Second,
	})

	cases := []struct {
		name     string
		fileSize int
		wantMin  time.Duration
	}{
		{"small file uses minimum", 1024, 120 * time.Second},
		{"1MB file", 1024 * 1024, 120 * time.Second},          // 30s < 120s min
		{"10MB file", 10 * 1024 * 1024, 300 * time.Second},    // 10 * 30s = 300s
		{"100MB file", 100 * 1024 * 1024, 3000 * time.Second}, // 100 * 30s
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			timeout := ext.computeTimeout(tc.fileSize)
			if timeout < tc.wantMin {
				t.Fatalf("timeout %s < expected minimum %s", timeout, tc.wantMin)
			}
		})
	}
}

func TestAudioExtractor_DefaultConfig(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{})

	if ext.pythonBinary != defaultPythonBinary {
		t.Fatalf("pythonBinary = %q, want %q", ext.pythonBinary, defaultPythonBinary)
	}
	if ext.modelSize != defaultWhisperModel {
		t.Fatalf("modelSize = %q, want %q", ext.modelSize, defaultWhisperModel)
	}
	if ext.maxFileSize != defaultMaxAudioSize {
		t.Fatalf("maxFileSize = %d, want %d", ext.maxFileSize, defaultMaxAudioSize)
	}
	if ext.minTimeout != defaultMinAudioTimeout {
		t.Fatalf("minTimeout = %s, want %s", ext.minTimeout, defaultMinAudioTimeout)
	}
}

func TestAudioExtractor_CustomConfig(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{
		PythonBinary:    "/usr/local/bin/python3.11",
		ModelSize:       "large-v3",
		DefaultLanguage: "fr",
		MaxFileSize:     1024 * 1024,
		TimeoutPerMin:   60 * time.Second,
		MinTimeout:      60 * time.Second,
	})

	if ext.pythonBinary != "/usr/local/bin/python3.11" {
		t.Fatalf("pythonBinary = %q", ext.pythonBinary)
	}
	if ext.modelSize != "large-v3" {
		t.Fatalf("modelSize = %q", ext.modelSize)
	}
	if ext.defaultLang != "fr" {
		t.Fatalf("defaultLang = %q", ext.defaultLang)
	}
}

func TestFormatTranscription_EmptySegments(t *testing.T) {
	t.Parallel()
	result := formatTranscription(nil)
	if result != "" {
		t.Fatalf("expected empty string, got %q", result)
	}
}

func TestFormatTranscription_SingleSegment(t *testing.T) {
	t.Parallel()
	segments := []TranscriptionSegment{
		{Start: 0, End: 5, Text: "Hello world"},
	}
	result := formatTranscription(segments)
	if !strings.Contains(result, "[00:00") {
		t.Fatalf("expected timestamp marker, got %q", result)
	}
	if !strings.Contains(result, "Hello world") {
		t.Fatalf("expected text content, got %q", result)
	}
}

func TestFormatTranscription_ParagraphBreaks(t *testing.T) {
	t.Parallel()
	segments := []TranscriptionSegment{
		{Start: 0, End: 10, Text: "First segment."},
		{Start: 10, End: 20, Text: "Second segment."},
		{Start: 35, End: 45, Text: "Third segment after break."},
		{Start: 45, End: 55, Text: "Fourth segment."},
	}
	result := formatTranscription(segments)

	// Should have paragraph break between second and third segments.
	parts := strings.Split(result, "\n\n")
	if len(parts) < 2 {
		t.Fatalf("expected at least 2 paragraphs, got %d: %q", len(parts), result)
	}
	if !strings.Contains(parts[0], "First segment.") {
		t.Fatalf("first paragraph missing first segment text")
	}
	if !strings.Contains(result, "Third segment after break.") {
		t.Fatalf("missing third segment text")
	}
}

func TestFormatTranscription_TimestampFormat(t *testing.T) {
	t.Parallel()
	segments := []TranscriptionSegment{
		{Start: 65, End: 90, Text: "One minute five."},
	}
	result := formatTranscription(segments)
	if !strings.Contains(result, "[01:05") {
		t.Fatalf("expected [01:05 timestamp, got %q", result)
	}
}

func TestFormatTimestamp(t *testing.T) {
	t.Parallel()
	cases := []struct {
		seconds float64
		want    string
	}{
		{0, "[00:00"},
		{30, "[00:30"},
		{60, "[01:00"},
		{65, "[01:05"},
		{3600, "[60:00"},
		{3661, "[61:01"},
	}
	for _, tc := range cases {
		t.Run(fmt.Sprintf("%.0fs", tc.seconds), func(t *testing.T) {
			t.Parallel()
			got := formatTimestamp(tc.seconds)
			if got != tc.want {
				t.Fatalf("formatTimestamp(%f) = %q, want %q", tc.seconds, got, tc.want)
			}
		})
	}
}

func TestExtensionFromContentType(t *testing.T) {
	t.Parallel()
	cases := []struct {
		contentType string
		fileName    string
		want        string
	}{
		{"audio/mpeg", "", ".mp3"},
		{"audio/wav", "", ".wav"},
		{"audio/x-wav", "", ".wav"},
		{"audio/flac", "", ".flac"},
		{"audio/mp4", "", ".m4a"},
		{"audio/ogg", "", ".ogg"},
		{"audio/aac", "", ".aac"},
		{"audio/opus", "", ".opus"},
		{"audio/x-ms-wma", "", ".wma"},
		{"unknown/type", "", ".wav"},
		{"audio/mpeg", "recording.mp3", ".mp3"},
		{"audio/mpeg", "recording.m4a", ".m4a"},
	}
	for _, tc := range cases {
		t.Run(tc.contentType+"_"+tc.fileName, func(t *testing.T) {
			t.Parallel()
			got := extensionFromContentType(tc.contentType, tc.fileName)
			if got != tc.want {
				t.Fatalf("extensionFromContentType(%q, %q) = %q, want %q",
					tc.contentType, tc.fileName, got, tc.want)
			}
		})
	}
}

func TestAudioExtractor_Extract_SilentAudio(t *testing.T) {
	t.Parallel()

	segments := []TranscriptionSegment{}
	text := formatTranscription(segments)
	if text != "" {
		t.Fatalf("expected empty text for silent audio, got %q", text)
	}
}

func TestAudioExtractor_InterfaceCompliance(t *testing.T) {
	t.Parallel()
	var _ Extractor = (*AudioExtractor)(nil)
}

func TestAudioExtractor_TempFileCleanup(t *testing.T) {
	t.Parallel()

	ext := NewAudioExtractor(AudioExtractorConfig{
		PythonBinary: "/nonexistent/python3",
		MinTimeout:   5 * time.Second,
	})

	input := []byte("fake audio data")
	_, _ = ext.Extract(context.Background(), input, ExtractOptions{
		ContentType: "audio/wav",
	})

	// Verify no leaked temp files by checking for our pattern.
	matches, err := filepath.Glob(filepath.Join(os.TempDir(), "memory-audio-*"))
	if err != nil {
		t.Fatalf("glob error: %v", err)
	}
	for _, m := range matches {
		info, err := os.Stat(m)
		if err != nil {
			continue
		}
		if time.Since(info.ModTime()) < 5*time.Second {
			t.Errorf("leaked temp file detected: %s", m)
		}
	}
}

func TestAudioExtractor_ExtractStream_FileTooLarge(t *testing.T) {
	t.Parallel()
	ext := NewAudioExtractor(AudioExtractorConfig{
		MaxFileSize: 50,
	})

	largeData := strings.NewReader(strings.Repeat("x", 100))
	_, err := ext.ExtractStream(context.Background(), largeData, ExtractOptions{
		ContentType: "audio/wav",
	})
	if err == nil {
		t.Fatal("expected error for oversized stream")
	}
	if !strings.Contains(err.Error(), "exceeds maximum") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestWhisperOutputParsing(t *testing.T) {
	t.Parallel()

	jsonStr := `{
		"language": "fr",
		"language_probability": 0.95,
		"duration": 65.3,
		"segments": [
			{"start": 0.0, "end": 30.0, "text": "Bonjour le monde"},
			{"start": 30.0, "end": 65.3, "text": "Comment allez-vous"}
		]
	}`

	var output whisperOutput
	if err := json.Unmarshal([]byte(jsonStr), &output); err != nil {
		t.Fatalf("parsing JSON: %v", err)
	}

	if output.Language != "fr" {
		t.Fatalf("language = %q, want %q", output.Language, "fr")
	}
	if output.Duration != 65.3 {
		t.Fatalf("duration = %f, want 65.3", output.Duration)
	}
	if len(output.Segments) != 2 {
		t.Fatalf("segments count = %d, want 2", len(output.Segments))
	}

	text := formatTranscription(output.Segments)
	if !strings.Contains(text, "Bonjour le monde") {
		t.Fatalf("missing first segment text in %q", text)
	}
	if !strings.Contains(text, "Comment allez-vous") {
		t.Fatalf("missing second segment text in %q", text)
	}
}

func TestWhisperOutputParsing_WithWordTimestamps(t *testing.T) {
	t.Parallel()

	jsonStr := `{
		"language": "en",
		"language_probability": 0.99,
		"duration": 5.0,
		"segments": [
			{
				"start": 0.0,
				"end": 5.0,
				"text": "Hello world",
				"words": [
					{"start": 0.0, "end": 0.5, "word": "Hello", "probability": 0.99},
					{"start": 0.6, "end": 1.0, "word": "world", "probability": 0.97}
				]
			}
		]
	}`

	var output whisperOutput
	if err := json.Unmarshal([]byte(jsonStr), &output); err != nil {
		t.Fatalf("parsing JSON: %v", err)
	}

	if len(output.Segments[0].Words) != 2 {
		t.Fatalf("words count = %d, want 2", len(output.Segments[0].Words))
	}
	if output.Segments[0].Words[0].Word != "Hello" {
		t.Fatalf("first word = %q, want %q", output.Segments[0].Words[0].Word, "Hello")
	}
	if output.Segments[0].Words[1].Probability != 0.97 {
		t.Fatalf("second word probability = %f, want 0.97", output.Segments[0].Words[1].Probability)
	}
}

func TestAudioExtractor_RegisterWithRegistry(t *testing.T) {
	t.Parallel()
	registry := NewExtractorRegistry()
	ext := NewAudioExtractor(AudioExtractorConfig{})
	registry.Register(ext)

	result, err := registry.Extract(context.Background(), []byte{}, ExtractOptions{
		ContentType: "audio/mpeg",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Reason != "empty audio input" {
		t.Fatalf("expected 'empty audio input' reason, got %q", result.Reason)
	}
}

File: sdks/ts/memory/src/ingest/audio.test.ts
// SPDX-License-Identifier: Apache-2.0

import { mkdtemp, writeFile, rm } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { Readable } from 'node:stream'
import { describe, expect, it, vi } from 'vitest'
import {
  createAudioExtractor,
  formatTranscription,
  type AudioExtractorConfig,
  type TranscriptionSegment,
} from './audio.js'
import { createExtractorRegistry } from './extractor.js'

/**
 * Creates a temporary shell script that simulates a faster-whisper
 * subprocess. The script ignores all arguments and outputs according
 * to the specified mode. Returns the script path and a cleanup function.
 */
const createMockScript = async (
  mode: 'success' | 'language_detect' | 'corrupt' | 'timeout' | 'availability_ok' | 'availability_fail',
): Promise<{ scriptPath: string; cleanup: () => Promise<void> }> => {
  const dir = await mkdtemp(join(tmpdir(), 'audio-test-'))

  const scripts: Record<string, string> = {
    success: `#!/bin/sh
echo '{"language":"en","language_probability":0.98,"duration":15.5,"segments":[{"start":0,"end":5.0,"text":"Hello world from the test.","words":[{"start":0,"end":0.5,"word":"Hello","probability":0.99},{"start":0.6,"end":1.0,"word":"world","probability":0.97}]},{"start":5.0,"end":15.5,"text":"This is a longer segment for testing."}]}'
`,
    language_detect: `#!/bin/sh
echo '{"language":"fr","language_probability":0.92,"duration":10.0,"segments":[{"start":0,"end":10.0,"text":"Bonjour le monde."}]}'
`,
    corrupt: `#!/bin/sh
echo "Error: could not decode audio file" >&2
exit 1
`,
    timeout: `#!/bin/sh
sleep 60
`,
    availability_ok: `#!/bin/sh
echo "ok"
`,
    availability_fail: `#!/bin/sh
echo "ModuleNotFoundError: No module named 'faster_whisper'" >&2
exit 1
`,
  }

  const scriptPath = join(dir, 'mock-python.sh')
  await writeFile(scriptPath, scripts[mode], { mode: 0o755 })

  return {
    scriptPath,
    cleanup: () => rm(dir, { recursive: true, force: true }),
  }
}

describe('createAudioExtractor', () => {
  const ext = createAudioExtractor()

  it('has the correct name', () => {
    expect(ext.name).toBe('audio-whisper')
  })

  it('declares audio content types', () => {
    expect(ext.contentTypes).toContain('audio/mpeg')
    expect(ext.contentTypes).toContain('audio/wav')
    expect(ext.contentTypes).toContain('audio/x-wav')
    expect(ext.contentTypes).toContain('audio/flac')
    expect(ext.contentTypes).toContain('audio/mp4')
    expect(ext.contentTypes).toContain('audio/ogg')
    expect(ext.contentTypes).toContain('audio/aac')
    expect(ext.contentTypes).toContain('audio/x-ms-wma')
    expect(ext.contentTypes).toContain('audio/opus')
  })

  it('reports capability with extensions and MIME types', () => {
    const cap = ext.capability()
    expect(cap.extensions).toContain('.mp3')
    expect(cap.extensions).toContain('.wav')
    expect(cap.extensions).toContain('.flac')
    expect(cap.extensions).toContain('.m4a')
    expect(cap.extensions).toContain('.ogg')
    expect(cap.extensions).toContain('.aac')
    expect(cap.requiresBinary).toBe(true)
    expect(cap.magicBytes.length).toBeGreaterThan(0)
  })

  it('reports MIME types in capability matching contentTypes', () => {
    const cap = ext.capability()
    for (const ct of ext.contentTypes) {
      expect(cap.mimeTypes).toContain(ct)
    }
  })
})

describe('createAudioExtractor - availability', () => {
  it('returns false when python binary does not exist', async () => {
    const ext = createAudioExtractor({ pythonBinary: '/nonexistent/python3' })
    const available = await ext.available()
    expect(available).toBe(false)
  })

  it('caches availability result on repeated calls', async () => {
    const ext = createAudioExtractor({ pythonBinary: '/nonexistent/python3' })
    const first = await ext.available()
    const second = await ext.available()
    expect(first).toBe(second)
    expect(first).toBe(false)
  })
})

describe('createAudioExtractor - extract', () => {
  it('returns skipped for empty input', async () => {
    const ext = createAudioExtractor()
    const result = await ext.extract(Buffer.alloc(0), { contentType: 'audio/wav' })
    expect(result.skipped).toBe(true)
    expect(result.reason).toBe('empty audio input')
    expect(result.text).toBe('')
  })

  it('throws for file exceeding max size', async () => {
    const ext = createAudioExtractor({ maxFileSizeBytes: 100 })
    const largeBuffer = Buffer.alloc(200)
    await expect(
      ext.extract(largeBuffer, { contentType: 'audio/wav' }),
    ).rejects.toThrow('exceeds maximum')
  })

  it('throws for stream exceeding max size', async () => {
    const ext = createAudioExtractor({ maxFileSizeBytes: 50 })
    const source = Readable.from([Buffer.alloc(100)])
    await expect(
      ext.extractStream(source, { contentType: 'audio/wav' }),
    ).rejects.toThrow('exceeds maximum')
  })
})

describe('createAudioExtractor - custom config', () => {
  it('accepts custom python binary path', () => {
    const ext = createAudioExtractor({ pythonBinary: '/usr/local/bin/python3.11' })
    expect(ext.name).toBe('audio-whisper')
  })

  it('accepts custom model size', () => {
    const ext = createAudioExtractor({ modelSize: 'large-v3' })
    expect(ext.name).toBe('audio-whisper')
  })

  it('accepts custom language', () => {
    const ext = createAudioExtractor({ defaultLanguage: 'fr' })
    expect(ext.name).toBe('audio-whisper')
  })

  it('accepts custom timeouts', () => {
    const ext = createAudioExtractor({
      timeoutPerMinuteMs: 60_000,
      minTimeoutMs: 60_000,
    })
    expect(ext.name).toBe('audio-whisper')
  })
})

describe('formatTranscription', () => {
  it('returns empty string for empty segments', () => {
    expect(formatTranscription([])).toBe('')
  })

  it('formats a single segment with timestamp', () => {
    const segments: TranscriptionSegment[] = [
      { start: 0, end: 5, text: 'Hello world' },
    ]
    const result = formatTranscription(segments)
    expect(result).toContain('[00:00')
    expect(result).toContain('Hello world')
  })

  it('inserts paragraph breaks every 30 seconds', () => {
    const segments: TranscriptionSegment[] = [
      { start: 0, end: 10, text: 'First segment.' },
      { start: 10, end: 20, text: 'Second segment.' },
      { start: 35, end: 45, text: 'Third segment after break.' },
      { start: 45, end: 55, text: 'Fourth segment.' },
    ]
    const result = formatTranscription(segments)
    const parts = result.split('\n\n')
    expect(parts.length).toBeGreaterThanOrEqual(2)
    expect(parts[0]).toContain('First segment.')
    expect(result).toContain('Third segment after break.')
  })

  it('formats timestamps correctly for times over a minute', () => {
    const segments: TranscriptionSegment[] = [
      { start: 65, end: 90, text: 'One minute five.' },
    ]
    const result = formatTranscription(segments)
    expect(result).toContain('[01:05')
  })

  it('formats timestamps for hour-long content', () => {
    const segments: TranscriptionSegment[] = [
      { start: 3661, end: 3690, text: 'Over an hour in.' },
    ]
    const result = formatTranscription(segments)
    expect(result).toContain('[61:01')
  })

  it('handles multiple segments within same paragraph', () => {
    const segments: TranscriptionSegment[] = [
      { start: 0, end: 5, text: 'Hello' },
      { start: 5, end: 10, text: 'world' },
      { start: 10, end: 15, text: 'test' },
    ]
    const result = formatTranscription(segments)
    // All within 30s window, should be one paragraph.
    expect(result.split('\n\n')).toHaveLength(1)
    expect(result).toContain('Hello')
    expect(result).toContain('world')
    expect(result).toContain('test')
  })
})

describe('whisper output parsing', () => {
  it('parses well-formed whisper JSON', () => {
    const jsonStr = JSON.stringify({
      language: 'fr',
      language_probability: 0.95,
      duration: 65.3,
      segments: [
        { start: 0.0, end: 30.0, text: 'Bonjour le monde' },
        { start: 30.0, end: 65.3, text: 'Comment allez-vous' },
      ],
    })

    const output = JSON.parse(jsonStr) as {
      language: string
      language_probability: number
      duration: number
      segments: TranscriptionSegment[]
    }
    expect(output.language).toBe('fr')
    expect(output.duration).toBe(65.3)
    expect(output.segments).toHaveLength(2)

    const text = formatTranscription(output.segments)
    expect(text).toContain('Bonjour le monde')
    expect(text).toContain('Comment allez-vous')
  })

  it('parses word-level timestamps', () => {
    const output = {
      language: 'en',
      language_probability: 0.99,
      duration: 5.0,
      segments: [
        {
          start: 0.0,
          end: 5.0,
          text: 'Hello world',
          words: [
            { start: 0.0, end: 0.5, word: 'Hello', probability: 0.99 },
            { start: 0.6, end: 1.0, word: 'world', probability: 0.97 },
          ],
        },
      ],
    }

    expect(output.segments[0]!.words).toHaveLength(2)
    expect(output.segments[0]!.words![0]!.word).toBe('Hello')
    expect(output.segments[0]!.words![1]!.probability).toBe(0.97)
  })
})

describe('createAudioExtractor - registry integration', () => {
  it('registers with extractor registry and routes audio types', async () => {
    const registry = createExtractorRegistry()
    const ext = createAudioExtractor()
    registry.register(ext)

    // Empty input should be handled (skipped), not unsupported.
    const result = await registry.extract(Buffer.alloc(0), {
      contentType: 'audio/mpeg',
    })
    expect(result.reason).toBe('empty audio input')
  })

  it('routes audio/wav to audio extractor', async () => {
    const registry = createExtractorRegistry()
    const ext = createAudioExtractor()
    registry.register(ext)

    const result = await registry.extract(Buffer.alloc(0), {
      contentType: 'audio/wav',
    })
    expect(result.reason).toBe('empty audio input')
  })

  it('routes audio/ogg to audio extractor', async () => {
    const registry = createExtractorRegistry()
    const ext = createAudioExtractor()
    registry.register(ext)

    const result = await registry.extract(Buffer.alloc(0), {
      contentType: 'audio/ogg',
    })
    expect(result.reason).toBe('empty audio input')
  })
})

describe('createAudioExtractor - subprocess mocking', () => {
  it('transcribes successfully with mocked subprocess', async () => {
    const { scriptPath, cleanup } = await createMockScript('success')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 30_000,
      })

      const input = Buffer.from('fake audio data for testing')
      const result = await ext.extract(input, { contentType: 'audio/wav' })

      expect(result.skipped).toBe(false)
      expect(result.text).toContain('Hello world from the test.')
      expect(result.language).toBe('en')
      expect(result.confidence).toBeGreaterThanOrEqual(0.9)
      expect(result.metadata.detected_language).toBe('en')
      expect(result.metadata.transcription_engine).toBe('faster-whisper')
      expect(result.metadata.segment_count).toBe('2')
      expect(result.metadata.segments).toBeTruthy()
    } finally {
      await cleanup()
    }
  })

  it('auto-detects language when no default is set', async () => {
    const { scriptPath, cleanup } = await createMockScript('language_detect')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 30_000,
      })

      const input = Buffer.from('fake french audio data')
      const result = await ext.extract(input, { contentType: 'audio/mp4' })

      expect(result.language).toBe('fr')
      expect(result.text).toContain('Bonjour le monde.')
    } finally {
      await cleanup()
    }
  })

  it('returns error for corrupt audio (non-zero exit)', async () => {
    const { scriptPath, cleanup } = await createMockScript('corrupt')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 30_000,
      })

      const input = Buffer.from('corrupt audio bytes')
      await expect(
        ext.extract(input, { contentType: 'audio/wav' }),
      ).rejects.toThrow('transcription failed')
    } finally {
      await cleanup()
    }
  })

  it('times out for a hanging subprocess', async () => {
    const { scriptPath, cleanup } = await createMockScript('timeout')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 1_500,
        timeoutPerMinuteMs: 500,
      })

      const input = Buffer.from('audio data that will time out')
      await expect(
        ext.extract(input, { contentType: 'audio/wav' }),
      ).rejects.toThrow('timed out')
    } finally {
      await cleanup()
    }
  })

  it('cleans up temp files after successful extraction', async () => {
    const { scriptPath, cleanup } = await createMockScript('success')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 30_000,
      })

      const input = Buffer.from('fake audio data for cleanup test')
      await ext.extract(input, { contentType: 'audio/wav' })

      // Temp files are cleaned up in a finally block; verify by checking
      // that no memory-audio-* directories remain in tmpdir that were
      // created in the last 5 seconds.
      const { readdirSync, statSync } = await import('node:fs')
      const tempBase = tmpdir()
      const entries = readdirSync(tempBase).filter(e => e.startsWith('memory-audio-'))
      const recent = entries.filter(e => {
        try {
          const info = statSync(join(tempBase, e))
          return Date.now() - info.mtimeMs < 5000
        } catch {
          return false
        }
      })
      expect(recent).toHaveLength(0)
    } finally {
      await cleanup()
    }
  })
})

describe('createAudioExtractor - abort signal / cancellation', () => {
  it('aborts extraction when signal is triggered', async () => {
    const { scriptPath, cleanup } = await createMockScript('timeout')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 60_000, // Long timeout so abort wins.
      })

      const controller = new AbortController()
      // Abort after a short delay.
      setTimeout(() => controller.abort(), 500)

      const input = Buffer.from('audio data to be aborted')
      await expect(
        ext.extract(input, { contentType: 'audio/wav' }, controller.signal),
      ).rejects.toThrow()
    } finally {
      await cleanup()
    }
  })

  it('aborts stream extraction when signal is triggered', async () => {
    const { scriptPath, cleanup } = await createMockScript('timeout')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        minTimeoutMs: 60_000,
      })

      const controller = new AbortController()
      setTimeout(() => controller.abort(), 500)

      const source = Readable.from([Buffer.from('audio stream data to abort')])
      await expect(
        ext.extractStream(source, { contentType: 'audio/wav' }, controller.signal),
      ).rejects.toThrow()
    } finally {
      await cleanup()
    }
  })
})

describe('createAudioExtractor - logging', () => {
  it('logs warning when faster-whisper is unavailable', async () => {
    const warn = vi.fn()
    const ext = createAudioExtractor({
      pythonBinary: '/nonexistent/python3',
      logger: { debug: vi.fn(), info: vi.fn(), warn, error: vi.fn() },
    })

    await ext.available()

    expect(warn).toHaveBeenCalledWith(
      'audio-whisper: faster-whisper not available',
      expect.objectContaining({ python: '/nonexistent/python3' }),
    )
  })

  it('does not log when faster-whisper is available', async () => {
    const { scriptPath, cleanup } = await createMockScript('availability_ok')
    try {
      const warn = vi.fn()
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
        logger: { debug: vi.fn(), info: vi.fn(), warn, error: vi.fn() },
      })

      const avail = await ext.available()

      expect(avail).toBe(true)
      expect(warn).not.toHaveBeenCalled()
    } finally {
      await cleanup()
    }
  })
})

describe('createAudioExtractor - availability race condition', () => {
  it('deduplicates concurrent availability checks', async () => {
    const { scriptPath, cleanup } = await createMockScript('availability_ok')
    try {
      const ext = createAudioExtractor({
        pythonBinary: scriptPath,
      })

      // Fire multiple concurrent calls. If the pending-promise pattern
      // works, they should all resolve to the same value without spawning
      // separate subprocesses.
      const results = await Promise.all([
        ext.available(),
        ext.available(),
        ext.available(),
      ])

      expect(results).toEqual([true, true, true])
    } finally {
      await cleanup()
    }
  })
})
